### ISSUE TITLE
objcet detection on  Android  is label The correct value cannot be used

### ISSUE THREAD

**Issue by baihualinxin** [tensorflow/models#5718]
------------------------

### System information
- **What is the top-level directory of the model you are using**:models-master\research\object_detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:ubuntu 16.04 windwos 10
- **TensorFlow installed from (source or binary)**:pip
- **TensorFlow version (use command below)**:1.10.0
- **Bazel version (if compiling from source)**:0.17.01
- **CUDA/cuDNN version**:cuda 9 cudnn 8
- **GPU model and memory**:GTX 1080Ti
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:
b'v1.10.0-rc1-19-g656e7a2b34' 1.10.0

### Describe the problem
Android demo in run Mobile phone label Fixed value
![792162767640418762](https://user-images.githubusercontent.com/13042809/48185098-fae4b380-e36e-11e8-99dc-1be27065de96.jpg)
![251355634106298905](https://user-images.githubusercontent.com/13042809/48185101-fc15e080-e36e-11e8-93e7-b933b7777d27.jpg)

label is  Identifying anything is 0.73

data set: klitti data
config:ssd_mobilenet_v1_quantized_300x300_coco14_sync.config
pbtxt：kitti_label_map.pbtxt
train：models-master\research\object_detection\legacy\train.py
        --logtostderr \
        --train_dir=path/to/train_dir \
        --pipeline_config_path=pipeline_config.pb

ckpt turn pb

models-master\research\object_detection\export_tflite_ssd_graph.py

python object_detection/export_tflite_ssd_graph \
    --pipeline_config_path path/to/pipeline.config \
    --trained_checkpoint_prefix path/to/model.ckpt-20000 \
    --output_directory path/to/exported_model_directory

pb turn tflite

bazel run --config=opt tensorflow/contrib/lite/toco:toco -- \
--input_file=$OUTPUT_DIR/tflite_graph.pb \
--output_file=$OUTPUT_DIR/detect.tflite \
--input_shapes=1,300,300,3 \
--input_arrays=normalized_input_image_tensor \
--output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3' \
--inference_type=QUANTIZED_UINT8 \
--mean_values=128 \
--std_values=128 \
--change_concat_input_ranges=false \
--allow_custom_ops

Explain:
mobilenet_v1_quantized_300x300_coco14_sync.config parameter
model {
  ssd {
    inplace_batchnorm_update: true
    freeze_batchnorm: false
    num_classes: 2
    box_coder {
      faster_rcnn_box_coder {
        y_scale: 10.0
        x_scale: 10.0
        height_scale: 5.0
        width_scale: 5.0
      }
    }
    matcher {
      argmax_matcher {
        matched_threshold: 0.5
        unmatched_threshold: 0.5
        ignore_thresholds: false
        negatives_lower_than_unmatched: true
        force_match_for_each_row: true
        use_matmul_gather: true
      }
    }
    similarity_calculator {
      iou_similarity {
      }
    }
    encode_background_as_zeros: true
    anchor_generator {
      ssd_anchor_generator {
        num_layers: 6
        min_scale: 0.2
        max_scale: 0.95
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.3333
      }
    }
    image_resizer {
      fixed_shape_resizer {
        height: 300
        width: 300
      }
    }
    box_predictor {
      convolutional_box_predictor {
        min_depth: 0
        max_depth: 0
        num_layers_before_predictor: 0
        use_dropout: false
        dropout_keep_probability: 0.8
        kernel_size: 1
        box_code_size: 4
        apply_sigmoid_to_scores: false
        class_prediction_bias_init: -4.6
        conv_hyperparams {
          activation: RELU_6,
          regularizer {
            l2_regularizer {
              weight: 0.00004
            }
          }
          initializer {
            random_normal_initializer {
              stddev: 0.01
              mean: 0.0
            }
          }
          batch_norm {
            scale: true,
            center: true,
            decay: 0.97,
            epsilon: 0.001,
          }
        }
      }
    }
    feature_extractor {
      type: 'ssd_mobilenet_v1'
      min_depth: 16
      depth_multiplier: 1.0
      conv_hyperparams {
        activation: RELU_6,
        regularizer {
          l2_regularizer {
            weight: 0.00004
          }
        }
        initializer {
          random_normal_initializer {
            stddev: 0.01
            mean: 0.0
          }
        }
        batch_norm {
          scale: true,
          center: true,
          decay: 0.97,
          epsilon: 0.001,
        }
      }
      override_base_feature_extractor_hyperparams: true
    }
    loss {
      classification_loss {
        weighted_sigmoid_focal {
          alpha: 0.75,
          gamma: 2.0
        }
      }
      localization_loss {
        weighted_smooth_l1 {
        }
      }
      classification_weight: 1.0
      localization_weight: 1.0
    }
    normalize_loss_by_num_matches: true
    normalize_loc_loss_by_codesize: true
    post_processing {
      batch_non_max_suppression {
        score_threshold: 1e-8
        iou_threshold: 0.6
        max_detections_per_class: 100
        max_total_detections: 100
      }
      score_converter: SIGMOID
    }
  }
}

train_config: {
  #fine_tune_checkpoint: "PATH_TO_BE_CONFIGURED/model.ckpt"
  batch_size: 32
  sync_replicas: true
  startup_delay_steps: 0
  replicas_to_aggregate: 8
  num_steps: 50000
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  #data_augmentation_options {
   # ssd_random_crop {
   # }
  #}
  optimizer {
    momentum_optimizer: {
      learning_rate: {
        cosine_decay_learning_rate {
          learning_rate_base: .2
          total_steps: 50000
          warmup_learning_rate: 0.06
          warmup_steps: 2000
        }
      }
      momentum_optimizer_value: 0.9
    }
    use_moving_average: false
  }
  max_number_of_boxes: 100
  unpad_groundtruth_tensors: false
}
train_input_reader: {
  tf_record_input_reader {
    input_path: "D:/SSD_KITTI/data/test_data/kitti_train.tfrecord"
  }
  label_map_path: "G:/KITTI_record/tfrecord/kitti_label_map.pbtxt"
}

eval_config: {
  metrics_set: "coco_detection_metrics"
  use_moving_averages: false
  num_examples: 8000
}

eval_input_reader: {
  tf_record_input_reader {
    input_path: "D:/SSD_KITTI/data/test_data/kitti_val.tfrecord"
  }
  label_map_path: "G:/KITTI_record/tfrecord/kitti_label_map.pbtxt"
  shuffle: false
  num_readers: 1
}

graph_rewriter {
  quantization {
    delay: 48000
    activation_bits: 8
    weight_bits: 8
  }
}

problem:
1.config parameter  Is there a problem?

2.Is it necessary to modify?
export_tflite_ssd_graph.py 
 flags.DEFINE_integer('max_detections', 10,
                     'Maximum number of detections (boxes) to show.')
flags.DEFINE_integer('max_classes_per_detection', 1,
                     'Number of classes to display per detection box.')

3.I don't know what else is wrong？
Or is there a problem with the model?

4.Android demo 
  private static final String TF_OD_API_MODEL_FILE = "detect.tflite";
  private static final String TF_OD_API_LABELS_FILE = "file:///android_asset/labels_list.txt";

What parameters do I need to change?

**Comment by baihualinxin** [comment]
@nealwu @bmabey @moonboots @alextp

**Comment by haikuoyao** [comment]
I'm on the same boat.
I wonder if it's because of training.  
I just use the normal way to train.
I didn't add any quantized thing in training, for instance, 
```
    tf.contrib.quantize.create_training_graph()
```

**Comment by baihualinxin** [comment]
@haikuoyao 
I am a self-demarcated data XML format, and convert CVS, CVS to tfrecords. There's no problem with training to turn tflite into an Android demo

**Comment by baihualinxin** [comment]
@nealwu What's the problem

**Comment by anjosma** [comment]
I'm facing this problem too. Someone?

**Comment by SteveIb** [comment]
I'm trying to detect more than 10 objects in the image ( which is default )
I'm usin the following commands:
bazel run -c opt tensorflow/contrib/lite/toco:toco -- --input_file=$OUTPUT_DIR/tflite_graph.pb --output_file=$OUTPUT_DIR/mobile_net_500.tflite --input_shapes=1,300,300,3 --input_arrays=normalized_input_image_tensor --output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3' --inference_type=FLOAT --max_detections=500 --max_classes_per_detection=1 --allow_custom_ops 

I also modified 
export_tflite_ssd_graph.py
flags.DEFINE_integer('max_detections', 500 <--- instead of 10,
'Maximum number of detections (boxes) to show.')
flags.DEFINE_integer('max_classes_per_detection', 1,
'Number of classes to display per detection box.')

but still giving 10 objects as output in the android [1,10,4].

any idea?

**Comment by baihualinxin** [comment]
@SteveIb 
bazel run -c opt tensorflow/contrib/lite/toco:toco 

I have tried or reported an error.

Toco is best used for slim's model, not for ssd_mobilenet

**Comment by baihualinxin** [comment]
@anjosma 
I have done the test 

 create_kitti_tf_record.py Script problems

**Comment by baihualinxin** [comment]
@haikuoyao 
I don't think that's the problem

**Comment by baihualinxin** [comment]
@nealwu 
Please give a solution to this problem

**Comment by SteveIb** [comment]
the solution which worked for me(to change number of objects):

1-change number of objects to detect inside export_tflite_ssd_graph.py
flags.DEFINE_integer('max_detections', 5000,
                     'Maximum number of detections (boxes) to show.')
I changed it to 5000.

2- ckpt turn pb
models-master\research\object_detection\export_tflite_ssd_graph.py

then you will have two files  - tflite_graph.pbtxt
 - tflite_graph.pb

3-use toco to convert it to tflite either float or quantized, not the size should be 300*300

bazel run -c opt tensorflow/contrib/lite/toco:toco -- 
--input_file=/home/dell/TFObject-detection/tensorflow/models/research/object_detection/OutputSSDtflite/tflite_graph.pb 
--output_file=/home/dell/TFObject-detection/tensorflow/models/research/object_detection/OutputSSDtflite/ssdtflite.tflite
--input_shapes=1,300,300,3 
--input_arrays=normalized_input_image_tensor 
--output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3'  
--inference_type=FLOAT 
--allow_custom_ops



===Regarding your problem 0.73, maybe a hint could be useful:
it might be that you need to change in your config file

batch_non_max_suppression {
score_threshold: 1e-8
iou_threshold: 0.6
max_detections_per_class: 100
max_total_detections: 100
} 

see this example from export_ssd_tflite .py

Example Usage (in which we change the NMS iou_threshold to be 0.5 and
NMS score_threshold to be 0.0):
python object_detection/export_tflite_ssd_graph \
    --pipeline_config_path path/to/ssd_mobilenet.config \
    --trained_checkpoint_prefix path/to/model.ckpt \
    --output_directory path/to/exported_model_directory
    --config_override " \
            model{ \
            ssd{ \
              post_processing { \
                batch_non_max_suppression { \
                        score_threshold: 0.0 \
                        iou_threshold: 0.5 \
                } \
             } \
          } \
       } \
       "

**Comment by baihualinxin** [comment]
@SteveIb 
I don't think it's a problem of 0.73， I've had 0.95
I do not know tfrecords，
It has been verified that I transferred to tfrecords

**Comment by baihualinxin** [comment]
@SteveIb 
I will verify the method you provided

**Comment by dunglt2015** [comment]
> the solution which worked for me(to change number of objects):
> 
> 1-change number of objects to detect inside export_tflite_ssd_graph.py
> flags.DEFINE_integer('max_detections', 5000,
> 'Maximum number of detections (boxes) to show.')
> I changed it to 5000.
> 
> 2- ckpt turn pb
> models-master\research\object_detection\export_tflite_ssd_graph.py
> 
> then you will have two files - tflite_graph.pbtxt
> 
> * tflite_graph.pb
> 
> 3-use toco to convert it to tflite either float or quantized, not the size should be 300*300
> 
> bazel run -c opt tensorflow/contrib/lite/toco:toco --
> --input_file=/home/dell/TFObject-detection/tensorflow/models/research/object_detection/OutputSSDtflite/tflite_graph.pb
> --output_file=/home/dell/TFObject-detection/tensorflow/models/research/object_detection/OutputSSDtflite/ssdtflite.tflite
> --input_shapes=1,300,300,3
> --input_arrays=normalized_input_image_tensor
> --output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3'
> --inference_type=FLOAT
> --allow_custom_ops
> } 
> } 
> } 
> } 
> "

@SteveIb I followed your solution to change number of objects but still get [1,10,4] output :(

**Comment by thusinh1969** [comment]
There is a TF_OD..._QUANTIZE flag in the DetectionActivity code file. You can try to turn it off to see if it works.

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
